### DISCUSSION TITLE
Win-SSHFS-Mounter (alternative SSHFS GUI/Systray app for Windows)

### DISCUSSION THREAD

**Discussion by 59de44955ebd** [winfsp/winfsp#556] [Show and tell]
Hi,
I just wanted to inform about this little Windows app I created recently:
https://github.com/59de44955ebd/Win-SSHFS-Mounter

It's similar to "SSHFS-Win Manager", but not based on Electron (i.e. yet another full chromium engine), and therefor much smaller in size. It's written in Python, but uses native Windows controls, no Qt etc. involved. Extra features are that it allows to import SSH/SFTP site configurations from FileZilla and WinSCP (including passwords), and that while importing it can convert PuTTY private keys (.ppk) to OpenSSH keys on the fly.

**Comment by billziss-gh** [comment]
Excellent work. Thank you for sharing.

**Reply by 59de44955ebd** [reply]
Dear Bill, thank you for the compliment. I know this is the wrong place, but if you don't mind, I have a small question that you might be able to answer: I tried to add an "Eject" item to the Explorer context menu which allows to disconnect SSHFS drives also direcftly from Explorer, see attached screenshot (Win 11 german). For this I implemented a commandline switch "Win-SSHFS-Mounter.exe -eject {drive letter}" and added a corresponding registry key to HKEY_CURRENT_USER\Software\Classes\Drive\shell. Win-SSHFS-Mounter.exe is a forced single instance exe, and when a new instance is started, it just forwards its commandline to the existing instance (via a WM_COPYDATA windows message) and quits.
So this works fine, but the ugly thing is that this context menu item obviously is now shown for all drives, instead of only for "ejectable" SSHFS drives. Are you aware of any method to create a specific registry key that will only be shown for SSHFS drives?

![explorer](https://github.com/winfsp/winfsp/assets/12278038/078f3747-4dec-4d72-ad3b-2f9c7c54df9f)

**Reply by billziss-gh** [reply]
I do not remember the details, but I believe I had to solve a similar problem in the [WinSpd](https://github.com/winfsp/winspd/tree/master) project.

This project include a context menu entry that can be used to eject the drive. The relevant C++ code is in [`ejectmenu.hpp`](https://github.com/winfsp/winspd/blob/master/src/shellex/ejectmnu.hpp), likely in the `QueryContextMenu` which decides whether to add the context menu entry for "eject" or not.

So unfortunately this solution would require you to write a shell extension.

**Reply by 59de44955ebd** [reply]
Nevermind, I just found a solution! Keys under HKEY_CURRENT_USER\Software\Classes\Drive\shell support an extra string value named "AppliesTo", and its value can be an OR combination of multiple drive letters, so e.g. "F: OR H: OR Z:". So whenever a SSHFS drive is mounted or unmounted, I can update this value accordingly, which makes sure that my "Eject SSHFS Drive" Explorer menu item now is only shown for actual SSHFS drives. If no drive is mounted at all, I set it to ":::", which should never match any existing drive (I hope).

**Reply by 59de44955ebd** [reply]
Sorry, I didn't refresh the open browser tab when I wrote my last reply, and therefor I hadn't seen your answer. Thank you! That's also what I had expected, it would surely be a cleaner solution than my current workaround, but I'm just too lazy to write a shell extension, so I guess I stick with what I have now. Thanks again!

**Reply by 59de44955ebd** [reply]
Just a quick follow-up note, latest release v0.3 now has this discussed feature of disconnecting/unmounting drives via Explorer context menu, and in addition to FileZilla and WinSCP it can now also import site configurations from PuTTY, KiTTY and Cyberduck:
https://github.com/59de44955ebd/Win-SSHFS-Mounter/releases

**Comment by 59de44955ebd** [comment]
Slightly OT, since this disussion is about WinFsp, but since its basically the same app, I thought I mention it anyway: in the meantime I also "ported" (well, sort of) this application to macOS, The macOS version is called "Mac-SSHFS-Mounter", requires macFUSE instead of WinFsp and uses the System Menu instead of the Windows System Tray:
https://github.com/59de44955ebd/Mac-SSHFS-Mounter/releases
